cmd/k-license/k-license.go

```diff
@@ -0,0 +1,199 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"github.com/spf13/cobra"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var excludeDirsLocations = "external/bazel_tools,.git,node_modules,_output,third_party,vendor,verify/boilerplate/test"
+var GENERATED_GO_MARKERS = [7]string{
+	"// Code generated by client-gen. DO NOT EDIT.",
+	"// Code generated by controller-gen. DO NOT EDIT.",
+	"// Code generated by counterfeiter. DO NOT EDIT.",
+	"// Code generated by deepcopy-gen. DO NOT EDIT.",
+	"// Code generated by informer-gen. DO NOT EDIT.",
+	"// Code generated by lister-gen. DO NOT EDIT.",
+	"// Code generated by protoc-gen-go. DO NOT EDIT.",
+}
+
+var codeFileExts = map[string]bool{
+	".go":    true,
+	".c":     true,
+	".h":     true,
+	".ipynb": true,
+	".py":    true,
+	".java":  true,
+	".cpp":   true,
+	".sh":    true,
+}
+
+var buildFileExts = map[string]bool{
+	"Makefile":   true,
+	"Dockerfile": true,
+}
+
+type Options struct {
+	templatesDir string
+	excludeDirs  string
+	path         string
+}
+type templateFileType struct {
+	fileExtension    string // store file extension strings like ".sh", "Makefile", etc.
+	templateFileName string // store template file names like "boilerplate.sh.txt", etc
+}
+
+var templateFileTypes = []templateFileType{
+	{".sh", "boilerplate.sh.txt"},
+	{"Makefile", "boilerplate.Makefile.txt"},
+	{"Dockerfile", "boilerplate.Dockerfile.txt"},
+	{".py", "boilerplate.py.txt"},
+	{".go", "boilerplate.go.txt"},
+}
+
+var Opts = &Options{}
+
+func main() {
+	rootCmd := &cobra.Command{
+		Use:   "k-license",
+		Short: "Tool for Adding license Headers",
+	}
+	addCmd := &cobra.Command{
+		Use:   "add",
+		Short: "Add Headers to files",
+		RunE: func(cmd *cobra.Command, args []string) error {
+
+			return Opts.Run()
+		},
+	}
+
+	addCmd.Flags().StringVar(&Opts.templatesDir, "templates", "../../hack/boilerplate", "directory containing license templates")
+	addCmd.Flags().StringVar(&Opts.excludeDirs, "exclude", excludeDirsLocations, "comma-separated list of directories to exclude")
+	addCmd.Flags().StringVar(&Opts.path, "path", ".", "Defaults to Current directory")
+	rootCmd.AddCommand(addCmd)
+	if err := rootCmd.Execute(); err != nil {
+		os.Exit(1)
+	}
+
+}
+func (opts *Options) Run() error {
+	excludeList := strings.Split(opts.excludeDirs, ",")
+	files := 0
+	err := filepath.Walk(opts.path, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() && contains(excludeList, info.Name()) {
```

Comment:
I am a little unsure about this. `info.Name()` is going to return only the base name of the dir. In a repo there can be multiple paths with the same base name. I don't think this would work for that.

I can't think of a straightforward way to handle this, but a possible way that could handle more cases like this, but still not all, could be:
Calling
```go
contains(excludeList, path)
```
and `contains` can be something like:
```go
func contains(exclude []string, path string) bool {
  for _, item := range exclude {
    var short, long string
    if len(path) < len(item) {
      short = path
      long = item
    } else {
      // other way around
    }
    if strings.HasSuffix(long, short) { return true }
  }
  return false
}
```
Instead of `HasSuffix`, it might be worth considering `filepath.Rel` too.

Reply:
> In a repo there can be multiple paths with the same base name. I don't think this would work for that.

One quick example from say -  kubernetes/test-infra:

Following files are on two different paths in the repo with the same base name:
- https://github.com/kubernetes/test-infra/blob/master/images/kubekins-e2e/Dockerfile
- https://github.com/kubernetes/test-infra/blob/master/images/builder/Dockerfile

@MadhavJivrajani could you please fill in more (or correct if I understood it wrong). Thank you! :slightly_smiling_face:

Reply:
I've checked it test infra repo seems to be working fine